wagtail_review/models.py

```diff
@@ -442,6 +441,43 @@ def on_action(self, task_state, user, action_name, reviewer=None, comment='', **
             review_token = Token(Reviewer.objects.get_or_create(internal=user)[0], task_state.page_revision, task_state)
             return redirect(get_review_url(review_token))
 
+    def get_task_states_user_can_moderate(self, user, **kwargs):
+        if self.is_reviewer_for_task(user) or user.is_superuser:
+            return TaskState.objects.filter(status=TaskState.STATUS_IN_PROGRESS, task=self.task_ptr)
+        else:
+            return TaskState.objects.none()
+
+
+class ReviewTask(ReviewMixin, Task):
+    reviewers = models.ManyToManyField(Reviewer)
+
+    panels = Task.panels + [ReviewerChooserPanel('reviewers')]
+
+    task_state_class = ReviewTaskState
+
+    def is_reviewer_for_task(self, user, reviewer=None):
+        if not reviewer:
+            reviewer, created = Reviewer.objects.get_or_create(internal=user)
+        return self.reviewers.filter(pk=reviewer.pk).exists()
+
     class Meta:
         verbose_name = _('Review task')
         verbose_name_plural = _('Review tasks')
+
+
+class GroupReviewTask(ReviewMixin, Task):
+    groups = models.ManyToManyField(Group, verbose_name=_('groups'), help_text=_('Pages at this step in a workflow will be commented on or approved by these groups of users'))
+
+    panels = Task.panels + [FieldPanel('groups', heading=_("Choose review groups"))]
+    exclude_on_edit = {'groups'}
+
+    task_state_class = ReviewTaskState
+
+    def is_reviewer_for_task(self, user, reviewer=None):
+        if reviewer and (not user or not user.is_authenticated):
+            user = get_user_model().objects.get(pk=reviewer.internal.pk)
```

Comment:
Up to this point, I thought that `user` was a required argument (and required to be an authenticated user), and `reviewer` was an optional extra argument to save a redundant database lookup. Now this bit of code has thrown me - it seems that there's some non-obvious business rules about when each argument comes into play, and I'm not sure if that relates to external reviewers or something else. Please can these rules be made explicit (e.g. with docstrings)? Without them, it's impossible whether all the other code here is correctly accounting for all the permutations of internal vs external, authenticated vs unauthenticated.

Reply:
Good point, I should make this clearer! Essentially, this is to deal with Wagtail Review's frontend, requests from which do not have an authenticated user attached, and in the case of an external reviewer, might not have a user to find. This is why I've been trying to write get_actions such that it can deal with being passed either a reviewer or a user (from frontend or backend respectively).

Reply:
I've added additional docstrings explaining the methods which take reviewer or user - hopefully this clarifies things.